orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/cf/CloudFoundryMonitorKatoServicesTaskTest.groovy

```diff
@@ -0,0 +1,157 @@
+/*
+ *  Copyright 2019 Pivotal, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.tasks.providers.cf
+
+import com.netflix.spinnaker.orca.ExecutionStatus
+import com.netflix.spinnaker.orca.clouddriver.KatoService
+import com.netflix.spinnaker.orca.clouddriver.model.Task
+import com.netflix.spinnaker.orca.clouddriver.model.TaskId
+import com.netflix.spinnaker.orca.pipeline.model.Execution
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import rx.Observable
+import spock.lang.Specification
+import spock.lang.Subject
+import spock.lang.Unroll
+
+class CloudFoundryMonitorKatoServicesTaskTest extends Specification {
+  @Subject task = new CloudFoundryMonitorKatoServicesTask(null)
+
+  @Unroll("result is #expectedStatus if kato task is #katoStatus and resultObjects is #resultObjects")
+  def "result depends on Kato task status and result object content"() {
+    given:
+    task.kato = Stub(KatoService) {
+      lookupTask(taskIdString, true) >>
+        Observable.from(new Task(taskIdString, new Task.Status(completed: completed, failed: failed), resultObjects, []))
+    }
+
+    and:
+    def stage = new Stage(Execution.newPipeline("orca"), opType, context)
+
+    when:
+    def result = task.execute(stage)
+
+    then:
+    result.status == expectedStatus
+    result.context == expectedContext
+
+    where:
+    cloudProvider = "cloud"
+    taskIdString = "kato-task-id"
+    credentials = "cf-foundation"
+    region = "org > space"
+    serviceInstanceName = "service-instance-name"
+    taskId = new TaskId(taskIdString)
+    context = [
+      "cloudProvider": cloudProvider,
+      "kato.last.task.id": taskId,
+      "credentials": credentials,
+      "region": region
+    ]
+
+    completed | failed | opType           | resultObjects                                      | expectedStatus             | expectedContext
```

Comment:
I like a good data driven test, but this one is hard to read. Mostly because the expectedContext is large and has a lot of duplication between cases.

Reply:
👍